### PR TITLE
Get request

### PR DESCRIPTION
**Pull request by llanahp**

### CODE REVIEW

File: src/class/HTTPCall/HTTPCall.cpp [hunk 1]
```diff
@@ -154,6 +154,51 @@ int HTTPCall::getClientFd(void) const
 	return this->_client_fd;
 }
 
+Path HTTPCall::getPathServerDirectory(void) const
+{
+	Path serverPath(this->_virtual_server->getConf()->getRoot()->get());
```
Comment (arikhativa) [inline src/class/HTTPCall/HTTPCall.cpp:159]: this is not so simple.
the path might be the path inside Location.
e.g.
```
server {
  root /var/www/html
  
  location /upload {
    root /tmp
  }
}
```

Comment (arikhativa) [reply src/class/HTTPCall/HTTPCall.cpp:159]: i think it is best to move this logic to matcher namespace and have something like `Path matcher::rootToRequest(Server, Location, BasicHTTPRequest)`

inside, if the root is not in the location, then we take the one from server

File: src/class/HTTPCall/HTTPCall.cpp [hunk 2]
```diff
@@ -154,6 +154,51 @@ int HTTPCall::getClientFd(void) const
...
+	return (serverPath);
+}
+
+std::list<const IErrorPage *> HTTPCall::getErrorPages(void) const
+{
+	return this->_virtual_server->getErrorPages();
+}
+
+std::list<const ILocation *>::const_iterator HTTPCall::searchMatchLocation(void) const
```
Comment (arikhativa) [inline src/class/HTTPCall/HTTPCall.cpp:168]: i wrote this logic in matcher namespace

File: src/class/HTTPCall/HTTPCall.cpp [hunk 3]
```diff
@@ -154,6 +154,51 @@ int HTTPCall::getClientFd(void) const
...
+			if (it == locations.begin())
+				tmp = it;
+			else if ((*it)->getPath().get().length() > (*tmp)->getPath().get().length())
+				tmp = it;
+		}
+	}
+	return (tmp);
+}
+
+bool HTTPCall::isAutoIndexOn(void) const
```
Comment (arikhativa) [inline src/class/HTTPCall/HTTPCall.cpp:186]: i think HTTPCall should have the IServer and ILocation as private vars.
we can get them with matcher after we read the client's call.
no need to look for them every time

File: src/class/HTTPCall/HTTPCall.cpp [hunk 4]
```diff
@@ -154,6 +154,51 @@ int HTTPCall::getClientFd(void) const
...
+{
+	std::list<const ILocation *>::const_iterator tmp = searchMatchLocation();
+	if (tmp != this->_virtual_server->getConf()->getLocations().end())
+		return (*tmp)->isAutoIndexOn();
+	return (false);
+}
+
+bool HTTPCall::canUpload(void) const
```
Comment (arikhativa) [inline src/class/HTTPCall/HTTPCall.cpp:194]: same

File: src/code/HTTPRequestHandler/HTTPRequestHandler.cpp [hunk 5]
```diff
@@ -4,8 +4,46 @@
 
 void HTTPRequestHandler::GET(HTTPCall &request)
 {
-	(void)request;
-	request.setResponse("HTTP/1.1 200 OK\r\n\r\nYOU SENT A GET REQUEST");
+	try
+	{
+		BasicHTTPRequest requestRecived = request.getBasicRequest();
```
Comment (arikhativa) [inline src/code/HTTPRequestHandler/HTTPRequestHandler.cpp:9]: i think HTTPCall have it's own BasicHTTPRequest. no need to create here a new one

File: src/code/HTTPRequestHandler/HTTPRequestHandler.cpp [hunk 6]
```diff
@@ -4,8 +4,46 @@
...
+		requestRecived.parseRaw();
+		Path url(request.getPathServerDirectory().get() + requestRecived.getPath());
+		if (httprequesthandlerGET::isDirectory(url.get()) && request.isAutoIndexOn())
+		{
+			ResponseHeader response(HTTPStatusCode(HTTPStatusCode::OK), request.getErrorPages());
```
Comment (arikhativa) [inline src/code/HTTPRequestHandler/HTTPRequestHandler.cpp:14]: nice flow, maybe split into a static func isDirectoryListing()?

File: src/code/HTTPRequestHandler/HTTPRequestHandler.cpp [hunk 7]
```diff
@@ -4,8 +4,46 @@
...
+			response.setBody(httprequesthandlerGET::getFileContent(url.get(), response));
+			return (request.setResponse(response.getResponse()));
+		}
+		else
+		{
+			ResponseHeader response(HTTPStatusCode(HTTPStatusCode::NOT_FOUND), request.getErrorPages());
+			return (request.setResponse(response.getResponse()));
+		}
+	}
+	catch (const BasicHTTPRequest::Incomplete &e)
```
Comment (arikhativa) [inline src/code/HTTPRequestHandler/HTTPRequestHandler.cpp:32]: if it is incomple then we need to read more from client till timeout

File: src/code/HTTPRequestHandlerGET/HTTPRequestHandlerGET.cpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
+#include <HTTPRequestHandlerGET/HTTPRequestHandlerGET.hpp>
+
+bool httprequesthandlerGET::isFileExists(const std::string &path)
```
Comment (arikhativa) [inline src/code/HTTPRequestHandlerGET/HTTPRequestHandlerGET.cpp:3]: why not just use FileManager directly?

**Review by arikhativa — CHANGES_REQUESTED** [review]
it looks good. there are a few notes, but it has a good structure :)

**Review by dVaGaymer — APPROVED** [review]

File: src/code/matcher/matcher.cpp [hunk 9]
```diff
@@ -69,13 +69,48 @@ const IServerConf *matcher::requestToServer(const IConf *conf, const IListen *so
...
+		}
+		it++;
 	}
-	return NULL;
+	if (tmp == list.end())
+		return NULL;
+	return (*tmp);
+}
+
+Path matcher::rootToRequest(const Server *server, const BasicHTTPRequest &req)
```
Comment (arikhativa) [inline src/code/matcher/matcher.cpp:88]: the HTTPCall object should have both ServerConf and Location. we should use them here:
```
Path matcher::rootToRequest(const HTTPCall &call)
{
	const ILocation *loc = requestToLocation(server->getConf(), req);
	if (call.location != NULL)
	{
		return (Path(call.location->getRoot()->get()));
	}
	return (Path(call.server->getRoot()->get()));
}
```

File: src/code/matcher/matcher.cpp [hunk 10]
```diff
@@ -69,13 +69,48 @@ const IServerConf *matcher::requestToServer(const IConf *conf, const IListen *so
...
+{
+	const ILocation *loc = requestToLocation(server->getConf(), req);
+	if (loc != NULL)
+	{
+		return (Path(loc->getRoot()->get()));
+	}
+	return (Path(server->getConf()->getRoot()->get()));
+}
+
+bool matcher::isAutoIndexOnToRequest(const Server *server, const BasicHTTPRequest &req)
```
Comment (arikhativa) [inline src/code/matcher/matcher.cpp:98]: same

File: src/code/matcher/matcher.cpp [hunk 11]
```diff
@@ -69,13 +69,48 @@ const IServerConf *matcher::requestToServer(const IConf *conf, const IListen *so
...
+{
+	const ILocation *loc = requestToLocation(server->getConf(), req);
+	if (loc != NULL)
+	{
+		return (loc->isAutoIndexOn());
+	}
+	return (false);
 }
+
+bool matcher::canUploadToRequest(const Server *server, const BasicHTTPRequest &req)
```
Comment (arikhativa) [inline src/code/matcher/matcher.cpp:108]: same

File: src/code/matcher/matcher.cpp [hunk 12]
```diff
@@ -69,13 +69,48 @@ const IServerConf *matcher::requestToServer(const IConf *conf, const IListen *so
...
 	std::list<const ILocation *> list = s->getLocations();
+	std::list<const ILocation *>::const_iterator tmp = list.end();
 	std::list<const ILocation *>::const_iterator it = list.begin();
-
 	while (it != list.end())
 	{
-		if ((*it)->getPath().get() == req.getPath())
-			return *it;
-		++it;
+		if ((req.getPath().find((*it)->getPath().get()) == 0) || (req.getPath() == "" && (*it)->getPath().get() == "/"))
```
Comment (arikhativa) [inline src/code/matcher/matcher.cpp:76]: i dont understand the logic here.
these 2 "if"s are very confusing.
what are you trying to change? why `if ((*it)->getPath().get() == req.getPath())` does not work?

**Review by arikhativa — CHANGES_REQUESTED** [review]

**Review by arikhativa — APPROVED** [review]

**Pull request merged**